File: Advanced Programming Methods/ToyLanguage-CyclicBarrier/src/sample/model/CloseRFile.java
package sample.model;

import sample.utils.FileData;
import sample.utils.MyIDictionary;
import sample.utils.MyIHeap;
import sample.utils.exceptions.*;

import java.io.BufferedReader;
import java.io.IOException;

public class CloseRFile implements IStmt {
    private Exp exp_file_id;
    public CloseRFile(Exp exp_file_id){
        this.exp_file_id=exp_file_id;
    }
    @Override
    public PrgState execute(PrgState state) throws EmptyException, NotExistingException, DivideByZeroException, OperatorException {
        MyIDictionary<String, Integer> st=state.getSymTable();
        MyIHeap<Integer, Integer> h=state.getHeap();
        int id=exp_file_id.eval(st, h);
        FileData fd=state.getFileTable().get(id);
        BufferedReader bf=fd.getFileDescriptor();
        try {
            bf.close();
        } catch (IOException e) {
            throw new InterpretorException("Cannot close the file");
        }
        //st.remove(exp_file_id);
        state.getFileTable().remove(id);
        return null;
    }

    @Override
    public String toString(){
        return "Close File "+exp_file_id;
    }
}

File: Advanced Programming Methods/toyLanguage/toyLanguage/model/CloseRFile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace toyLanguage.model
{
    class CloseRFile:IStmt
    {
        private string expFileId;
        public CloseRFile(string e)
        {
            expFileId = e;
        }
        public string ExpFileId
        {
            get { return expFileId; }
            set { expFileId = value; }
        }

        public PrgState Execute(PrgState p)
        {
            IMyDictionary<string, int> symbolt = p.SymTable;
            int id;
            symbolt.TryGetValue(expFileId, out id);
            FileData fd;
            p.FileTable.TryGetValue(id, out fd);
            StreamReader sr = fd.FileDescriptor;
            sr.Close();
            p.FileTable.Remove(id);
            return p;
        }

        public override string ToString()
        {
            return "CloseRFile(" + expFileId + ")";
        }
    }
}
